### PR TITLE
adding safe_int_mm

### PR DESCRIPTION
**Pull request by HDCharles**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #22
* #21
* #20
* #18
* __->__ #17

Summary: this is intended as a wrapper for torch._int_mm which
has a number of characteristics that make it difficult to use/debug.
This wrapped version:
1) falls back to a cpu compatible path if ran on cpu
2) short circuits to normal torch._int_mm if it detects torch compilation
to avoid perf degredation,
3) error checks tensor dimensions to avoid regions where cublas errors
and uses fallback instead
4) calls contiguous when w is non contiguous to avoid cublas silently
returning nonsense data
5) calls contiguous when x has a combination of being non-contiguous &
dim 0 is not a multiple of 8 which avoids a cryptic cublas error
*note* the torch compiled short circuit path doesn't have any error
checking so if that path is taken with tensors whose inner
multiplication dimension is <= 16 then it will error

triton graph of safe_int_mm:
==== __compiled_fn_0 =====
 <eval_with_key>.2 class GraphModule(torch.nn.Module):
    def forward(self, L_x_int8_ : torch.Tensor, L_w_int8_ : torch.Tensor):
        l_x_int8_ = L_x_int8_
        l_w_int8_ = L_w_int8_

        # File: /fsx/users/cdhernandez/protoquant/ao_experimental/quant_primitives.py:24, code: return torch._int_mm(x_int8, w_int8)
        _int_mm = torch._int_mm(l_x_int8_, l_w_int8_);  l_x_int8_ = l_w_int8_ = None
        return (_int_mm,)

[2023-05-10 00:32:08,655] torch._dynamo.output_graph.__graph: [DEBUG] TRACED GRAPH
 __compiled_fn_0 <eval_with_key>.2 opcode         name       target                                                      args                    kwargs
-------------  ---------  ----------------------------------------------------------  ----------------------  --------
placeholder    l_x_int8_  L_x_int8_                                                   ()                      {}
placeholder    l_w_int8_  L_w_int8_                                                   ()                      {}
call_function  _int_mm    <built-in method _int_mm of type object at 0x7fe27c3c48a0>  (l_x_int8_, l_w_int8_)  {}
output         output     output                                                      ((_int_mm,),)           {}

Test Plan: python test_quant_primitives.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D45757849](https://our.internmc.facebook.com/intern/diff/D45757849)

### CODE REVIEW

**Review by cpuhrsch — APPROVED** [review]
Great! When importing to fbcode just make sure to also update the TARGET files so that these tests run.

File: ao_experimental/quant_primitives.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
+import torch
+from torch import _dynamo
+
+def safe_int_mm(x_int8: torch.Tensor, w_int8: torch.Tensor):
```
Comment (vkuzo) [inline ao_experimental/quant_primitives.py]: nit: can we be consistent with https://pytorch.org/docs/stable/generated/torch.mm.html in terms of argument names?  It's not always weight and bias.

File: ao_experimental/quant_primitives.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    Return:
+        out (Tensor, torch.int32): the result of the matmul with device matching that of the inputs
+    """
+
+    # torch.compile path
+    if torch._dynamo.is_compiling():
+        return torch._int_mm(x_int8, w_int8)
+    
+    # error checking for cublas path    
+    device_cpu = 'cpu' in [w_int8.device.type, x_int8.device.type]
```
Comment (vkuzo) [inline ao_experimental/quant_primitives.py]: we should ensure the device is the same or error out here, otherwise this would enable people to write code which will handle devices incorrectly, and this function will magically fix it

**Comment by vkuzo** [comment]
> calls contiguous when x has a combination of being non-contiguous &
dim 0 is not a multiple of 8 which avoids a cryptic cublas error
note the torch compiled short circuit path doesn't have any error
checking so if that path is taken with tensors whose inner
multiplication dimension is <= 16 then it will return nonsense data

yes, great catch, I also got bitten by this and figured this one out the hard way in my experiments.  Can we do a warn-once for this in the torch.compile path?  Doesn't have to be in this PR, but would be good to add a TODO.

**Comment by vkuzo** [comment]
@HDCharles , can you attach the logs of what the dynamo graph looks like in the dynamo path to the summary?  Not required for landing, but I think it's a great thing to have in the PR summary as this PR will be really important to int8 quant workflows, and people will need to understand what this is doing.

File: ao_experimental/quant_primitives.py (hunk 2 above)
Comment (HDCharles) [reply ao_experimental/quant_primitives.py]: will do

**Comment by HDCharles** [comment]
> @HDCharles , can you attach the logs of what the dynamo graph looks like in the dynamo path to the summary? Not required for landing, but I think it's a great thing to have in the PR summary as this PR will be really important to int8 quant workflows, and people will need to understand what this is doing.

yeah, will do

**Review by vkuzo — APPROVED** [review]
looks great, let's land it!

**Comment by HDCharles** [comment]
@HDCharles has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D45757849).

**Comment by HDCharles** [comment]
@HDCharles has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D45757849).

**Comment by HDCharles** [comment]
@HDCharles has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D45757849).

**Comment by HDCharles** [comment]
@HDCharles has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D45757849).
